File: leetcode-75/2095_delete_the_middle_node_of_a_linked_list.cpp
struct ListNode {
    int val;
    ListNode* next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode* next) : val(x), next(next) {}
};

class Solution {
   public:
    ListNode* deleteMiddle(ListNode* head) {
        if (!head || !head->next) return nullptr;
        ListNode* slow = head;
        ListNode* fast = head;
        ListNode* prev = head;

        while (fast && fast->next) {
            prev = slow;
            slow = slow->next;
            fast = fast->next;
            if (fast) fast = fast->next;
        }

        prev->next = slow->next;
        slow->next = nullptr;
        return head;
    }
};

File: leetcode-75/2095_delete_the_middle_node_of_a_linked_list.py
from typing import Optional


class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution:
    def deleteMiddle(self, head: Optional[ListNode]) -> Optional[ListNode]:
        dmyHead = ListNode(0, head)
        prev = dmyHead
        midNode = head
        fast = head

        while (fast and fast.next):
            prev = midNode
            midNode = midNode.next
            fast = fast.next.next

        prev.next = midNode.next
        midNode.next = None

        return dmyHead.next